packages/contracts/utils/etherscan.ts

```diff
@@ -77,3 +82,417 @@ export const runTaskWithRetry = async (
+  implementation: {name: string; address: string} | undefined;
+};
+export const collectProxyWithImplementation = (
+  deployments: AllDeployments
+): CollectedProxyWithImplementation[] => {
+  const proxies = fetchProxies(deployments);
+  const implementations = fetchImplementations(deployments);
+
+  // join based on the first part of the name
+  return proxies.map(proxy => {
+    const baseName = proxy.name.replace('Proxy', '');
+    return {
+      name: baseName,
+      proxy,
+      implementation: implementations.find(imp => imp.name.includes(baseName)),
+    };
+  });
+};
+
+/**
+ * @returns the block explorer chain config for the current network the hre is connected to.
+ * The chain config comes from the hardhat config or the user config, paired with the default
+ * chain details exported from the hardhat-verify package.
+ */
+export const getChainBlockExplorerDetails = (
+  hre: HardhatRuntimeEnvironment
+): ChainConfig => {
+  const currentChainId = hre.network.config.chainId;
+
+  if (!currentChainId) {
+    throw new Error('Could not get chainId from Hardhat Runtime Environment.');
+  }
+  // deep copy to avoid modifying the original array
+  const customChainsCopy = Object.assign([], hre.config.etherscan.customChains);
+  const currentChainConfig = [
+    // custom chains has higher precedence than builtin chains
+    ...customChainsCopy.reverse(), // the last entry has higher precedence
+    ...builtinChains,
+  ].find(({chainId}) => chainId === currentChainId);
+
+  if (!currentChainConfig) {
+    throw new Error(
+      `ChainId ${currentChainId} not supported. Please add the chain details to the hardhat config.`
+    );
+  }
+
+  return currentChainConfig || null;
+};
+
+/**
+ * @returns The Etherscan API key for the current network the hre is connected to.
+ * We look for this key in the hardhat config, but it must be set in the user's `.env` file.
+ */
+export const getApiKey = (
+  hre: HardhatRuntimeEnvironment,
+  chain: ChainConfig
+): string => {
+  // check the apiKey first in the hardhat config, then in the user config
+  const apiKeyByNetwork = (hre.config.etherscan.apiKey ??
+    hre.userConfig.etherscan?.apiKey) as Record<string, string>;
+
+  if (!apiKeyByNetwork) {
+    throw new Error(`Missing Etherscan API keys in the hardhat config.`);
+  }
+
+  if (!(chain.network in apiKeyByNetwork)) {
+    throw new Error(
+      `${chain.network} not supported. Please add the API key to the hardhat config.`
+    );
+  }
+  return apiKeyByNetwork[chain.network];
+};
+
+/**
+ * @returns The API endpoint to hit for the given chain to verify a proxy contract's implementation.
+ */
+export const generateExplorerIsThisAProxyURL = (
+  hre: HardhatRuntimeEnvironment
+): string => {
+  const chain = getChainBlockExplorerDetails(hre);
+  const apiKey = getApiKey(hre, chain);
+
+  const baseUrl = chain.urls.apiURL;
+
+  return (
+    baseUrl + `?module=contract&action=verifyproxycontract&apikey=${apiKey}`
```

Comment:
these concat.s are reliable?
is there an oficial solution?
if not, and this meant to be an internal solution, then i would not add it to the official deploy flow.
perhaps we can still have it but manually triggered by us internally ..etc